pkg/sfu/buffer/dependencydescriptorparser.go

```diff
@@ -0,0 +1,104 @@
+package buffer
+
+import (
+	"fmt"
+
+	dd "github.com/livekit/livekit-server/pkg/sfu/dependencydescriptor"
+
+	"github.com/livekit/protocol/logger"
+)
+
+type videoLayer struct {
+	spatial  int
+	temporal int
+}
+
+type DependencyDescriptorParser struct {
+	structure         *dd.FrameDependencyStructure
+	ddExt             uint8
+	logger            logger.Logger
+	onMaxLayerChanged func(int, int)
+	decodeTargetLayer []videoLayer
+}
+
+func NewDependencyDescriptorParser(ddExt uint8, logger logger.Logger, onMaxLayerChanged func(int, int)) *DependencyDescriptorParser {
+	logger.Infow("creating video stream receiver", "ddExt", ddExt)
+	return &DependencyDescriptorParser{
+		ddExt:             ddExt,
+		logger:            logger,
+		onMaxLayerChanged: onMaxLayerChanged,
+	}
+}
+
+func (r *DependencyDescriptorParser) Parse(ep *ExtPacket) error {
+	pkt := ep.Packet
+	if ddBuf := pkt.GetExtension(r.ddExt); ddBuf != nil {
+		var ddVal dd.DependencyDescriptor
+		ext := &dd.DependencyDescriptorExtension{
+			Descriptor: &ddVal,
+			Structure:  r.structure,
+		}
+		_, err := ext.Unmarshal(ddBuf)
+		if err != nil {
+			r.logger.Infow("failed to parse generic dependency descriptor", "err", err)
+			return err
+		}
+
+		if ddVal.AttachedStructure != nil && !ddVal.FirstPacketInFrame {
+			r.logger.Infow("ignoring non-first packet in frame with attached structure")
+			return nil
+		}
+
+		if ddVal.AttachedStructure != nil {
+			var maxSpatial, maxTemporal int
+			r.structure = ddVal.AttachedStructure
+			r.decodeTargetLayer = r.decodeTargetLayer[:0]
+			for target := 0; target < r.structure.NumDecodeTargets; target++ {
+				layer := videoLayer{0, 0}
+				for _, t := range r.structure.Templates {
+					if t.DecodeTargetIndications[target] != dd.DecodeTargetNotPresent {
+						if layer.spatial < t.SpatialId {
+							layer.spatial = t.SpatialId
+						}
+						if layer.temporal < t.TemporalId {
+							layer.temporal = t.TemporalId
+						}
+					}
+				}
+				if layer.spatial > maxSpatial {
+					maxSpatial = layer.spatial
+				}
+				if layer.temporal > maxTemporal {
+					maxTemporal = layer.temporal
+				}
+				r.decodeTargetLayer = append(r.decodeTargetLayer, layer)
+			}
+			r.logger.Debugw("max layer changed", "maxSpatial", maxSpatial, "maxTemporal", maxTemporal)
+			r.onMaxLayerChanged(maxSpatial, maxTemporal)
```

Comment:
Maybe make this callback a go routine or make it is a go routine where this handled. I am guessing this flows back to all subscribers and back to their stream allocators. The stream allocator posts the events to a channel to process, but if this happens in packet processing path, a large number of subscribers means it could be making a lot of calls in packet path.